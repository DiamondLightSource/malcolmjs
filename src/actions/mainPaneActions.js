/**
 * Created by twi18192 on 25/08/15.
 */

import AppDispatcher from '../dispatcher/appDispatcher.js';
import appConstants  from '../constants/appConstants.js';

class MainPaneActions {
  constructor()
    {

    }

  toggleFooter1(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.FOOTER_TOGGLE,
      item: item
    })
  }

  toggleConfigPanel(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.CONFIG_TOGGLE,
      item: item
    })
  }
  toggleFavPanel(item){
    AppDispatcher.handleViewAction({
      actionType: appConstants.FAV_TOGGLE,
      item: item
    })
  }
}

const mainPaneActions = new MainPaneActions();

export default mainPaneActions;
