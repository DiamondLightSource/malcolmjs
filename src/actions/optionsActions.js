/**
 * Created by Ian Gillingham 29 August 2017.
 */

import AppDispatcher from '../dispatcher/appDispatcher';
import appConstants from '../constants/appConstants';

let optionsActions = {

  GraphicsStyleSelect: function(gstyle){
    AppDispatcher.handleViewAction({
      actionType: appConstants.OPTIONS_ACTION_GSTYLE,
      item:gstyle
    })
  },
  RightPanePinnedSelect: function(rightPinned){
    AppDispatcher.handleViewAction({
      actionType: appConstants.OPTIONS_ACTION_RIGHT_PINNED,
      item: rightPinned
    })
  }
};

export default optionsActions;
