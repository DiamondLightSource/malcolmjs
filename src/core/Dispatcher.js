/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
