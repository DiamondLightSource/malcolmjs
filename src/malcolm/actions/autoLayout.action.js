import ELK from 'elkjs/lib/elk.bundled';
import { malcolmPutAction, malcolmSetFlag } from '../malcolmActionCreators';

const calculateHeight = (layoutEngine, mri) => {
  const zoomFactor = layoutEngine.diagramModel.zoom / 100;
  return (
    layoutEngine.getNodeDimensions(layoutEngine.diagramModel.nodes[mri])
      .height / zoomFactor
  );
};

const runAutoLayout = () => (dispatch, getState) => {
  const state = getState();
  const { blocks } = state.malcolm.layout;
  const { parentBlock, mainAttribute } = state.malcolm;
  const { links } = state.malcolm.layoutEngine.diagramModel;

  const graph = {
    id: 'root',
    layoutOptions: {
      'elk.algorithm': 'layered',
      separateConnectedComponents: true,
      'spacing.componentComponent': 70,
      spacing: 100,
      'spacing.nodeNodeBetweenLayers': 70,
      'elk.edgeRouting': 'SPLINES',
    },
    children: blocks.map(b => ({
      id: b.mri,
      width: 120,
      height: calculateHeight(state.malcolm.layoutEngine, b.mri),
      x: b.position.x,
      y: b.position.y,
      properties: {
        portConstraints: 'FIXED_ORDER',
      },
      ports: b.ports.map((p, i) => ({
        id: `${b.mri}-${p.label}`,
        properties: {
          side: p.input ? 'WEST' : 'EAST',
          index: p.input ? b.ports.length - i : i,
        },
        width: 12,
        height: 12,
      })),
    })),
    edges: Object.values(links).map((link, i) => ({
      id: `e${i}`,
      sources: [link.sourcePort.id],
      targets: [link.targetPort.id],
    })),
  };

  // The JSON graph can be visualised here https://rtsys.informatik.uni-kiel.de/elklive/json.html
  console.log(JSON.stringify(graph));
  console.log(
    `Graph can be visualised at https://rtsys.informatik.uni-kiel.de/elklive/json.html`
  );

  const elk = new ELK();
  return elk
    .layout(graph)
    .then(graphLayout => {
      console.log('completed auto layout');
      const centerX =
        graphLayout.children.reduce((prev, next) => prev + next.x, 0) /
        graphLayout.children.length;
      const centerY =
        graphLayout.children.reduce((prev, next) => prev + next.y, 0) /
        graphLayout.children.length;

      // we only need to send an update for the visible blocks
      const updatedLayout = {
        name: blocks.map(b => b.name),
        visible: blocks.map(b => b.visible),
        mri: blocks.map(b => b.mri),
        x: blocks
          .map(b => b.mri)
          .map(mri => graphLayout.children.find(c => c.id === mri).x - centerX),
        y: blocks
          .map(b => b.mri)
          .map(mri => graphLayout.children.find(c => c.id === mri).y - centerY),
      };

      dispatch(malcolmSetFlag([parentBlock, mainAttribute], 'pending', true));
      dispatch(
        malcolmPutAction([parentBlock, mainAttribute, 'value'], updatedLayout)
      );
    })
    .catch(err => {
      console.log('error creating auto layout');
      console.log(err);
    });
};

export default {
  runAutoLayout,
};
