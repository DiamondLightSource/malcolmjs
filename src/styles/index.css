@-moz-document url-prefix() {
    /* Make all native selects have text butting up to their start to make
     * them look like inputs
     * https://github.com/callemall/material-ui/issues/8946
     */
    select {
        text-indent: -2px;
    }
}

/* Self hosted fonts */
/* https://google-webfonts-helper.herokuapp.com/fonts/roboto?subsets=latin */
/* roboto-regular - latin */
@font-face {
    font-family: 'Roboto';
    font-style: normal;
    font-weight: 400;
    src: local('Roboto'), local('Roboto-Regular'),
    url('./roboto-v18-latin-regular.woff') format('woff'); /* Chrome 6+, Firefox 3.6+, IE 9+, Safari 5.1+ */
}

/* roboto-500 - latin */
@font-face {
    font-family: 'Roboto';
    font-style: normal;
    font-weight: 500;
    src: local('Roboto Medium'), local('Roboto-Medium'),
    url('./roboto-v18-latin-500.woff') format('woff'); /* Chrome 6+, Firefox 3.6+, IE 9+, Safari 5.1+ */
}

/* https://github.com/google/material-design-icons/blob/master/iconfont/MaterialIcons-Regular.woff */
@font-face {
    font-family: 'Material Icons';
    font-style: normal;
    font-weight: 400;
    src: local('Material Icons'),
    local('MaterialIcons-Regular'),
    url(./MaterialIcons-Regular-2.2.0.woff) format('woff'); /* Chrome 6+, Firefox 3.6+, IE 9+, Safari 5.1+ */
}

.material-icons {
    font-family: 'Material Icons';
    font-weight: normal;
    font-style: normal;
    font-size: 24px;  /* Preferred icon size */
    display: inline-block;
    line-height: 1;
    text-transform: none;
    letter-spacing: normal;
    word-wrap: normal;
    white-space: nowrap;
    direction: ltr;

    /* Support for all WebKit browsers. */
    -webkit-font-smoothing: antialiased;
    /* Support for Safari and Chrome. */
    text-rendering: optimizeLegibility;

    /* Support for Firefox. */
    -moz-osx-font-smoothing: grayscale;

    /* Support for IE. */
    font-feature-settings: 'liga';
}

