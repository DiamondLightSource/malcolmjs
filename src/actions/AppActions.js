/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Dispatcher from '../core/Dispatcher';
import ActionTypes from '../constants/ActionTypes';
import ExecutionEnvironment from 'react/lib/ExecutionEnvironment';
import http from 'superagent';

export default {

  navigateTo(path, options) {
    if (ExecutionEnvironment.canUseDOM) {
      if (options && options.replace) {
        window.history.replaceState({}, document.title, path);
      } else {
        window.history.pushState({}, document.title, path);
      }
    }

    Dispatcher.handleViewAction({
      actionType: ActionTypes.CHANGE_LOCATION,
      path
    });
  },

  loadPage(path, cb) {
    Dispatcher.handleViewAction({
      actionType: ActionTypes.LOAD_PAGE,
      path
    });

    http.get('/api/page' + path)
      .accept('application/json')
      .end((err, res) => {
        Dispatcher.handleServerAction({
          actionType: ActionTypes.LOAD_PAGE,
          path,
          err,
          page: res.body
        });
        if (cb) {
          cb();
        }
      });
  }

};
