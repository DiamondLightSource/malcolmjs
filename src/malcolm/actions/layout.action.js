import {
  MalcolmSelectPortType,
  MalcolmMakeBlockVisibleType,
  MalcolmShowBinType,
  MalcolmInLayoutDeleteZoneType,
  MalcolmResetPortsType,
} from '../malcolm.types';
import {
  malcolmPutAction,
  malcolmSetFlag,
  malcolmSelectBlock,
} from '../malcolmActionCreators';
import blockUtils from '../blockUtils';
import { snackbarState } from '../../viewState/viewState.actions';
import { idSeparator } from '../../layout/layout.component';
import { sinkPort } from '../malcolmConstants';

const findPort = (blocks, id) => {
  const path = id.split(idSeparator);
  const block = blocks.find(b => b.mri === path[0]);
  const port = block.ports.find(p => p.label === path[1]);

  return port;
};

const selectPortInClient = (portId, start) => ({
  type: MalcolmSelectPortType,
  payload: {
    portId,
    start,
  },
});

export const selectPort = (portId, start) => (dispatch, getState) => {
  if (!portId) {
    if (getState().malcolm.layoutState.startPortForLink) {
      dispatch({
        type: MalcolmResetPortsType,
        payload: {},
      });
    }
    return;
  }

  const port = findPort(getState().malcolm.layout.blocks, portId);

  if (port && port.input && port.value !== port.tag) {
    // Reset the ports
    dispatch({
      type: MalcolmResetPortsType,
      payload: {},
    });

    // Display an error message
    dispatch(
      snackbarState(
        true,
        `Error: Existing link to input port ${
          port.label
        }, delete existing link first.`
      )
    );
  } else {
    dispatch(selectPortInClient(portId, start));

    const { layoutState, layout } = getState().malcolm;
    const { startPortForLink, endPortForLink } = layoutState;

    if (startPortForLink && endPortForLink) {
      const startPort = findPort(layout.blocks, startPortForLink);
      const endPort = findPort(layout.blocks, endPortForLink);

      const outputPort = startPort.input ? endPort : startPort;

      // 2. update on the server
      const path = [
        ...(startPort.input ? startPortForLink : endPortForLink).split(
          idSeparator
        ),
        'value',
      ];

      dispatch(malcolmSetFlag(path, 'pending', true));
      dispatch(malcolmPutAction(path, outputPort.tag));
    }
  }
};

const makeBlockVisible = (mriList, positionList, visibleList) => (
  dispatch,
  getState
) => {
  let mri;
  let position;
  let visible;
  if (!(mriList instanceof Array)) {
    mri = [mriList];
    position = [positionList];
    visible = [visibleList === undefined ? true : visibleList];
  } else {
    mri = [...mriList];
    position = [...positionList];
    visible = [...visibleList];
  }

  const state = getState().malcolm;
  const positionRelativeToCenter = position.map(pos => ({
    x: pos.x - state.layoutState.layoutCenter.x,
    y: pos.y - state.layoutState.layoutCenter.y,
  }));

  mri.forEach((blockMri, index) => {
    dispatch({
      type: MalcolmMakeBlockVisibleType,
      payload: {
        mri: blockMri,
        position: positionRelativeToCenter[index],
        visible: visible[index],
      },
    });
  });

  const blockName = state.parentBlock;
  const attributeName = state.mainAttribute;
  const layoutAttribute = blockUtils.findAttribute(
    state.blocks,
    blockName,
    attributeName
  );

  if (layoutAttribute && mri.length > 0) {
    const visibleBlockIndices = mri.map(blockMri =>
      layoutAttribute.raw.value.mri.findIndex(val => val === blockMri)
    );
    const updateLayout = {
      name: visibleBlockIndices.map(
        index => layoutAttribute.raw.value.name[index]
      ),
      visible,
      mri,
      x: positionRelativeToCenter.map(pos => pos.x),
      y: positionRelativeToCenter.map(pos => pos.y),
    };

    dispatch(malcolmSetFlag([blockName, attributeName], 'pending', true));
    dispatch(
      malcolmPutAction([blockName, attributeName, 'value'], updateLayout)
    );
  }
};

const deleteBlocks = () => (dispatch, getState) => {
  const state = getState().malcolm;
  const { selectedBlocks } = state.layoutState;

  makeBlockVisible(
    selectedBlocks,
    selectedBlocks.map(() => state.layoutState.layoutCenter),
    selectedBlocks.map(() => false)
  )(dispatch, getState);
  selectedBlocks.forEach(b => {
    dispatch(malcolmSelectBlock(b, false));
  });
};

const deleteLinks = () => (dispatch, getState) => {
  const state = getState().malcolm;

  state.layoutState.selectedLinks.forEach(linkId => {
    const [, , blockMri, linkAttr] = linkId.split(idSeparator);
    const portAttribute = blockUtils.findAttribute(
      state.blocks,
      blockMri,
      linkAttr
    );
    if (portAttribute) {
      const portNullValue = portAttribute.raw.meta.tags
        .find(t => t.indexOf(sinkPort) > -1)
        .split(':')
        .slice(-1)[0];
      dispatch(malcolmSetFlag([blockMri, linkAttr], 'pending', true));
      dispatch(malcolmPutAction([blockMri, linkAttr], portNullValue));
    }
  });
};

const showLayoutBin = show => ({
  type: MalcolmShowBinType,
  payload: {
    visible: show,
  },
});

const mouseInsideDeleteZone = insideZone => ({
  type: MalcolmInLayoutDeleteZoneType,
  payload: {
    insideZone,
  },
});

export default {
  selectPort,
  makeBlockVisible,
  showLayoutBin,
  mouseInsideDeleteZone,
  deleteBlocks,
  deleteLinks,
};
