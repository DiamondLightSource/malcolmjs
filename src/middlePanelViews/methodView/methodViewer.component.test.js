import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles';
import { blue } from '@material-ui/core/colors/index';
import MethodViewer from './methodViewer.component';
import { malcolmUpdateMethodInput } from '../../malcolm/actions/method.actions';
import MockCircularBuffer from '../../malcolm/reducer/attribute.reducer.mocks';
import { malcolmTypes } from '../../malcolmWidgets/attributeDetails/attributeSelector/attributeSelector.component';

jest.mock('../../malcolm/actions/method.actions');

const mockTheme = createMuiTheme({
  palette: {
    type: 'dark',
    primary: blue,
  },
  alarmState: {
    warning: '#e6c01c',
    error: '#e8001f',
    disconnected: '#9d07bb',
  },
});

const pushTestData = freshArchive => {
  freshArchive.value.push({
    took: {
      first: '0',
      second: { someJSON: true },
      third: ['anArray'],
    },
    returned: { first: false, second: 'someText' },
  });
  freshArchive.timeStamp.push({
    runTime: new Date(0),
    returnTime: new Date(1536583618205),
  });
  freshArchive.alarmState.push(0);
  freshArchive.value.push({
    took: { first: 0 },
    returned: {},
  });
  freshArchive.timeStamp.push({
    runTime: new Date(-14258382000),
    returnTime: new Date(123456789),
  });
  freshArchive.alarmState.push(4);
  return freshArchive;
};

describe('Method viewer', () => {
  let shallow;
  let mockStore;
  let mount;
  let testInputs;
  let freshArchive;
  let testInputValues;

  const getItem = jest.fn();
  const setItem = jest.fn();

  const state = {
    malcolm: {
      blocks: { Test: { attributes: [] } },
      blockArchive: { Test: { attributes: [] } },
    },
  };

  const testOutputs = {
    first: {
      label: 'First output',
      tags: ['widget:led'],
    },
    second: {
      label: 'Second output',
      tags: ['widget:textupdate'],
    },
    third: {
      label: 'Third output',
      tags: ['widget:textupdate'],
      typeid: malcolmTypes.stringArray,
    },
  };

  const testOutputValues = {
    first: true,
    second: 'done',
    third: ['something finished'],
  };

  const testMethodViewer = (
    outputValues,
    selected,
    errorMsg,
    archiveContents
  ) => {
    const inputState = {};
    Object.keys(testInputValues).forEach(input => {
      inputState[input] =
        testInputValues[input].value === undefined
          ? { value: testInputValues[input] }
          : testInputValues[input];
    });
    const outputState = {};
    Object.keys(outputValues).forEach(output => {
      outputState[output] = { value: outputValues[output] };
    });
    const method = {
      calculated: {
        name: 'Method',
        path: ['Test', 'Method'],
        errorState: !!errorMsg,
        errorMessage: errorMsg,
        inputs: inputState,
        outputs: outputState,
        isMethod: true,
      },
      raw: {
        meta: {
          label: 'Test',
          defaults: {},
          takes: { elements: { ...testInputs } },
          returns: { elements: { ...testOutputs } },
        },
      },
    };

    state.malcolm.blocks.Test.attributes = [method];
    state.malcolm.blockArchive.Test.attributes = [archiveContents];

    return (
      <MuiThemeProvider theme={mockTheme}>
        <MethodViewer
          store={mockStore(state)}
          blockName="Test"
          attributeName="Method"
          subElement={selected}
          classes={{}}
        />
      </MuiThemeProvider>
    );
  };

  beforeEach(() => {
    getItem.mockClear();
    setItem.mockClear();
    malcolmUpdateMethodInput.mockClear();

    mockStore = myState => ({
      getState: () => myState,
      subscribe: () => {},
      dispatch: () => {},
    });

    shallow = createShallow({ dive: true });
    mount = createMount();

    testInputValues = {
      first: 3,
      second: { value: 'four', isAString: true },
      third: {
        value: ['SEVEN'],
        flags: { rows: [] },
        meta: {
          typeid: malcolmTypes.stringArray,
          tags: ['widget:textinput'],
          writeable: true,
        },
      },
    };

    testInputs = {
      first: {
        label: 'First input',
        description: 'a test input',
        tags: ['widget:textinput'],
      },
      second: {
        label: 'Second input',
        description: 'a test input',
        tags: ['widget:tree'],
      },
      third: {
        label: 'Third input',
        description: 'an array',
        tags: ['widget:textinput'],
        typeid: malcolmTypes.stringArray,
        writeable: true,
      },
    };

    freshArchive = {
      name: 'Method',
      timeStamp: new MockCircularBuffer(5),
      value: new MockCircularBuffer(5),
      alarmState: new MockCircularBuffer(5),
    };
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders input correctly with no initial value', () => {
    testInputValues = {};
    const wrapper = shallow(
      testMethodViewer({}, ['takes', 'first'], undefined, freshArchive)
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders input correctly with initial value', () => {
    const wrapper = shallow(
      testMethodViewer({}, ['takes', 'first'], undefined, freshArchive)
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders output correctly with no initial value', () => {
    testInputValues = {};
    const wrapper = shallow(
      testMethodViewer({}, ['returns', 'first'], undefined, freshArchive)
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders output correctly with initial value', () => {
    testInputValues = {};
    const wrapper = shallow(
      testMethodViewer(
        testOutputValues,
        ['returns', 'first'],
        undefined,
        freshArchive
      )
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders output array correctly', () => {
    testInputValues = {};
    const wrapper = shallow(
      testMethodViewer(
        testOutputValues,
        ['returns', 'third'],
        undefined,
        freshArchive
      )
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders method param archive if selected parameter isnt a tree', () => {
    testInputValues = {};
    const wrapper = shallow(
      testMethodViewer({}, ['takes', 'first'], undefined, freshArchive)
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders top-level method param archive if no selected parameter', () => {
    testInputValues = {};
    const wrapper = shallow(
      testMethodViewer({}, undefined, undefined, pushTestData(freshArchive))
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  /* TODO: fix tests for virtualizedTable-based array
  it('copy params button hooks in correctly on top-level method param archive', () => {
    testInputValues = {};
    const wrapper = mount(
      testMethodViewer({}, undefined, undefined, pushTestData(freshArchive))
    );
    wrapper
      .find('button')
      .first()
      .simulate('click');
    expect(malcolmUpdateMethodInput).toHaveBeenCalledTimes(3);
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'third',
      ['anArray']
    );
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'second',
      { someJSON: true }
    );
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'first',
      '0'
    );
    malcolmUpdateMethodInput.mockClear();
    wrapper
      .find('button')
      .last()
      .simulate('click');
    expect(malcolmUpdateMethodInput).toHaveBeenCalledTimes(3);
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'third',
      undefined
    );
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'second',
      undefined
    );
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'first',
      0
    );
  });
  */
  // it('updates method input after 1s inactivity', () => {});

  it('renders correctly for array once local state is initialised', () => {
    const wrapper = shallow(
      testMethodViewer({}, ['takes', 'third'], undefined, freshArchive)
    );
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('add row hooks up correctly for array', () => {
    const wrapper = mount(
      testMethodViewer({}, ['takes', 'third'], undefined, freshArchive)
    );
    wrapper
      .find('button')
      .last()
      .simulate('click');
    expect(malcolmUpdateMethodInput).toHaveBeenCalledTimes(1);
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'third',
      ['SEVEN', '']
    );
  });
  /* TODO: fix tests for virtualizedTable-based array
  it('update input hooks up correctly for array', () => {
    const wrapper = mount(
      testMethodViewer({}, ['takes', 'third'], undefined, freshArchive)
    );
    console.log(wrapper.debug());
    wrapper
      .find('input')
      .first()
      .simulate('change', { target: { value: '180' } });
    expect(malcolmUpdateMethodInput).toHaveBeenCalledTimes(1);
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'third',
      ['180']
    );
  });

  it('update input hooks up correctly for array if writeable is false', () => {
    testInputs.third.writeable = false;
    const wrapper = mount(
      testMethodViewer({}, ['takes', 'third'], undefined, freshArchive)
    );
    wrapper
      .find('input')
      .first()
      .simulate('change', { target: { value: '180' } });
    expect(malcolmUpdateMethodInput).not.toHaveBeenCalled();
  });
  */
});
