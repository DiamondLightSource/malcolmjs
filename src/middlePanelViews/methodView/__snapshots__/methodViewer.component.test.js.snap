// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Method viewer renders correctly for array once local state is initialised 1`] = `
<WithStyles(WidgetTable)
  addRow={[Function]}
  attribute={
    Object {
      "calculated": Object {
        "errorMessage": undefined,
        "errorState": false,
        "inputs": Object {
          "first": Object {
            "value": 3,
          },
          "second": Object {
            "isAString": true,
            "value": "four",
          },
          "third": Object {
            "flags": Object {
              "rows": Array [],
            },
            "meta": Object {
              "tags": Array [
                "widget:textinput",
              ],
              "typeid": "malcolm:core/StringArrayMeta:1.0",
              "writeable": true,
            },
            "value": Array [
              "SEVEN",
            ],
          },
        },
        "isMethod": true,
        "name": "Method",
        "outputs": Object {},
        "path": Array [
          "Test",
          "Method",
        ],
      },
      "raw": Object {
        "defaults": Object {},
        "label": "Test",
        "returns": Object {
          "elements": Object {
            "first": Object {
              "label": "First output",
              "tags": Array [
                "widget:led",
              ],
            },
            "second": Object {
              "label": "Second output",
              "tags": Array [
                "widget:textupdate",
              ],
            },
            "third": Object {
              "label": "Third output",
              "tags": Array [
                "widget:textupdate",
              ],
              "typeid": "malcolm:core/StringArrayMeta:1.0",
            },
          },
        },
        "takes": Object {
          "elements": Object {
            "first": Object {
              "description": "a test input",
              "label": "First input",
              "tags": Array [
                "widget:textinput",
              ],
            },
            "second": Object {
              "description": "a test input",
              "label": "Second input",
              "tags": Array [
                "widget:tree",
              ],
            },
            "third": Object {
              "description": "an array",
              "label": "Third input",
              "tags": Array [
                "widget:textinput",
              ],
              "typeid": "malcolm:core/StringArrayMeta:1.0",
              "writeable": true,
            },
          },
        },
      },
    }
  }
  closePanelHandler={[Function]}
  eventHandler={[Function]}
  infoClickHandler={[Function]}
  localState={
    Object {
      "flags": Object {
        "rows": Array [],
      },
      "meta": Object {
        "tags": Array [
          "widget:textinput",
        ],
        "typeid": "malcolm:core/StringArrayMeta:1.0",
        "writeable": true,
      },
      "value": Array [
        "SEVEN",
      ],
    }
  }
  rowClickHandler={[Function]}
  setFlag={[Function]}
/>
`;

exports[`Method viewer renders input correctly with initial value 1`] = `
<MethodArchive
  defaultTab={0}
  methodArchive={
    Object {
      "alarmState": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "name": "Method",
      "timeStamp": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "value": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
    }
  }
  openPanels={
    Object {
      "child": undefined,
      "parent": undefined,
    }
  }
  selectedParam={
    Array [
      "takes",
      "first",
    ]
  }
/>
`;

exports[`Method viewer renders input correctly with no initial value 1`] = `
<MethodArchive
  defaultTab={0}
  methodArchive={
    Object {
      "alarmState": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "name": "Method",
      "timeStamp": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "value": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
    }
  }
  openPanels={
    Object {
      "child": undefined,
      "parent": undefined,
    }
  }
  selectedParam={
    Array [
      "takes",
      "first",
    ]
  }
/>
`;

exports[`Method viewer renders method param archive if selected parameter isnt a tree 1`] = `
<MethodArchive
  defaultTab={0}
  methodArchive={
    Object {
      "alarmState": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "name": "Method",
      "timeStamp": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "value": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
    }
  }
  openPanels={
    Object {
      "child": undefined,
      "parent": undefined,
    }
  }
  selectedParam={
    Array [
      "takes",
      "first",
    ]
  }
/>
`;

exports[`Method viewer renders output array correctly 1`] = `
<WithStyles(WidgetTable)
  addRow={[Function]}
  attribute={
    Object {
      "calculated": Object {
        "errorMessage": undefined,
        "errorState": false,
        "inputs": Object {},
        "isMethod": true,
        "name": "Method",
        "outputs": Object {
          "first": Object {
            "value": true,
          },
          "second": Object {
            "value": "done",
          },
          "third": Object {
            "value": Array [
              "something finished",
            ],
          },
        },
        "path": Array [
          "Test",
          "Method",
        ],
      },
      "raw": Object {
        "defaults": Object {},
        "label": "Test",
        "returns": Object {
          "elements": Object {
            "first": Object {
              "label": "First output",
              "tags": Array [
                "widget:led",
              ],
            },
            "second": Object {
              "label": "Second output",
              "tags": Array [
                "widget:textupdate",
              ],
            },
            "third": Object {
              "label": "Third output",
              "tags": Array [
                "widget:textupdate",
              ],
              "typeid": "malcolm:core/StringArrayMeta:1.0",
            },
          },
        },
        "takes": Object {
          "elements": Object {
            "first": Object {
              "description": "a test input",
              "label": "First input",
              "tags": Array [
                "widget:textinput",
              ],
            },
            "second": Object {
              "description": "a test input",
              "label": "Second input",
              "tags": Array [
                "widget:tree",
              ],
            },
            "third": Object {
              "description": "an array",
              "label": "Third input",
              "tags": Array [
                "widget:textinput",
              ],
              "typeid": "malcolm:core/StringArrayMeta:1.0",
              "writeable": true,
            },
          },
        },
      },
    }
  }
  closePanelHandler={[Function]}
  eventHandler={[Function]}
  infoClickHandler={[Function]}
  localState={
    Object {
      "flags": Object {
        "rows": Array [],
      },
      "meta": Object {
        "label": "Third output",
        "tags": Array [
          "widget:textupdate",
        ],
        "typeid": "malcolm:core/StringArrayMeta:1.0",
      },
      "value": Array [
        "something finished",
      ],
    }
  }
  rowClickHandler={[Function]}
  setFlag={[Function]}
/>
`;

exports[`Method viewer renders output correctly with initial value 1`] = `
<MethodArchive
  defaultTab={0}
  methodArchive={
    Object {
      "alarmState": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "name": "Method",
      "timeStamp": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
      "value": MockCircularBuffer [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ],
    }
  }
  openPanels={
    Object {
      "child": undefined,
      "parent": undefined,
    }
  }
  selectedParam={
    Array [
      "returns",
      "first",
    ]
  }
/>
`;

exports[`Method viewer renders output correctly with no initial value 1`] = `
<WithStyles(Typography)>
  No results yet...run method first!
</WithStyles(Typography)>
`;

exports[`Method viewer renders top-level method param archive if no selected parameter 1`] = `
<WithStyles(WidgetTable)
  addRow={[Function]}
  attribute={
    Object {
      "calculated": Object {},
      "raw": Object {
        "meta": Object {
          "elements": Object {
            "alarm": Object {
              "label": "Alarm state",
              "tags": Array [
                "info:alarm",
              ],
            },
            "copyParams": Object {
              "label": "Reuse run params",
              "tags": Array [
                "info:button",
              ],
              "writeable": true,
            },
            "postTime": Object {
              "label": "Time run",
              "tags": Array [
                "widget:textupdate",
              ],
            },
            "returnStatus": Object {
              "label": "Return Status",
              "tags": Array [
                "widget:textupdate",
              ],
            },
            "returnTime": Object {
              "label": "Time results received",
              "tags": Array [
                "widget:textupdate",
              ],
            },
          },
        },
        "value": Object {
          "alarm": Array [
            0,
            4,
          ],
          "copyParams": Array [
            Object {
              "action": [Function],
              "label": "Copy",
            },
            Object {
              "action": [Function],
              "label": "Copy",
            },
          ],
          "postTime": Array [
            1970-01-01T00:00:00.000Z,
            1969-07-19T23:20:18.000Z,
          ],
          "returnStatus": Array [
            undefined,
            undefined,
          ],
          "returnTime": Array [
            2018-09-10T12:46:58.205Z,
            1970-01-02T10:17:36.789Z,
          ],
        },
      },
    }
  }
  closePanelHandler={[Function]}
  eventHandler={[Function]}
  hideInfo={true}
  infoClickHandler={[Function]}
  rowClickHandler={[Function]}
  setFlag={[Function]}
/>
`;
